### PR TITLE
Write all voting data to db and sum the counts in Lorre

### LINKED ISSUE #253: Governance Tables
We want to keep track of Tezos's governance process in our data model. In order to do this, we will create three tables, "proposals", "bakers", and "ballots", as well as add fields to the blocks table.

The schema is attached, and all types and foreign key information will be in there, as well as information about what RPC calls to make in order to obtain the data required. Some notes:

Proposals: Will have the fields "protocol_hash", "block_hash", and "block_level", so we know what proposal is currently be considered during the pseudo "timestamp" that is identified by "block_hash" and "block_level". 

Bakers: Will have fields "pkh", "rolls", "block_hash", "block_level", so we know the account ids of bakers and how many rolls they have. 1 roll = 10k tz, is proportional to the amount of "voting power" the baker has. This is also pseudo "timestamped" by "block_hash" and "block_level". 

Ballots: Will have fields "pkh", "ballot, "block_hash", "block_level", so we know what votes a given user with account id "pkh" votes on a specific proposal. Since this is also pseudo "timestamped" by "block_hash" and "block_level", we can join into the blocks table to see what specific proposal this is for.

BlocksAdd:  We will add the fields "period_kind", "current_expected_quorum", and "active_proposal" to know what proposal is currently under consideration, if any, what section of the proposal review process we are currently in at the time of block creation, and the current_expected_quorum necessary to accept this proposal. The reason for current_expected_quorum is this value changes dynamically as more and more proposals are considered.

### PR DESCRIPTION
**Pull request by ivanopagano**
closes #253 

 * introduce data types for voting information
 * add fetching of voting details: proposals, ballots, bakers
 * add voting data tables
 * now slick generates a full case class for the operations table so
   remove the TableMapping object and all code using it, and use the
   slick-generated class instead
 * define a database conversion for voting proposals
 * refactor to remove ambiguities with proposals and ballot operations

## Required changes to DB schema
```sql
CREATE TABLE public.proposals (
    protocol_hash character varying NOT NULL,
    block_id character varying NOT NULL,
    block_level integer NOT NULL
);

CREATE TABLE public.bakers (
    pkh character varying NOT NULL,
    rolls integer NOT NULL,
    block_id character varying NOT NULL,
    block_level integer NOT NULL
);

CREATE TABLE public.ballots (
    pkh character varying NOT NULL,
    ballot character varying NOT NULL,
    block_id character varying NOT NULL,
    block_level integer NOT NULL
);

CREATE INDEX ix_proposals_protocol ON public.proposals USING btree (protocol_hash); 

ALTER TABLE ONLY public.proposals
    ADD CONSTRAINT proposal_block_id_fkey FOREIGN KEY (block_id) REFERENCES public.blocks(hash);

ALTER TABLE ONLY public.bakers
    ADD CONSTRAINT baker_block_id_fkey FOREIGN KEY (block_id) REFERENCES public.blocks(hash);

ALTER TABLE ONLY public.ballots
    ADD CONSTRAINT ballot_block_id_fkey FOREIGN KEY (block_id) REFERENCES public.blocks(hash);

```

## Required changes to Application configuration
None

### CODE REVIEW

**Comment by ivanopagano** [comment]
This is a PR to the base branch for work on voting data, it's not directly merging to `master`, so that you can review only what's relevant progress over the previous PR #267

**Comment by vishakh** [comment]
The coveralls check is failing. :(

**Comment by ivanopagano** [comment]
> The coveralls check is failing. :(

I'll add tests on new conversions that I have missed somehow.

Yet I'm afraid that coveralls won't increase steadily if we don't design a way to emulate nodes behavior for data-fetch testing

File: doc/conseil.sql [hunk 1]
```diff
@@ -328,6 +363,13 @@ CREATE INDEX ix_operations_source ON public.operations USING btree (source);
 
 CREATE INDEX ix_accounts_checkpoint_block_level ON public.accounts_checkpoint USING btree (block_level);
 
+--
+-- Name: ix_proposals_protocol; Type: INDEX; Schema: public; Owner: -
+--
+
+CREATE INDEX ix_proposals_protocol ON public.proposals USING btree (protocol_hash);
```
Comment (ba205) [inline doc/conseil.sql]: What inspired this index?

Comment (vishakh) [reply doc/conseil.sql]: The most common query pattern on this table will be to get proposal details by protocol hash.

File: src/main/scala/tech/cryptonomic/conseil/Lorre.scala [hunk 2]
```diff
@@ -191,6 +202,45 @@ object Lorre extends App with TezosErrors with LazyLogging with LorreAppConfig {
...
+        /* combineAll reduce List[Option[Int]] => Option[Int] by summing all ints present
+        * |+| is shorthand syntax to sum Option[Int] together using Int's sums
+         * Any None in the operands will make the whole operation collapse in a None result
+         */
+        for {
+          storedProposals <- db.run(writeProposal)
+          storedBakers <- db.run(writeBakers).map(_.combineAll)
+          storedBallots <-  db.run(writeBallots).map(_.combineAll)
+        } yield storedProposals |+| storedBakers |+| storedBallots
+    }
```
[the remark this answers is not in the document]
Comment (ivanopagano) [reply src/main/scala/tech/cryptonomic/conseil/Lorre.scala]: I will see how to put it all in

File: src/main/scala/tech/cryptonomic/conseil/tezos/TezosNodeFetchers.scala [hunk 3]
```diff
@@ -139,4 +139,55 @@ trait BlocksDataFetchers {
...
+
+    val makeUrl = (block: Block) => s"blocks/${block.data.hash.value}/votes/proposals"
+
+    override val fetchData =
+      Kleisli(blocks => node.runBatchedGetQuery(network, blocks, makeUrl, fetchConcurrency))
+
+    override val decodeData = Kleisli{
+      json => JsonDecoders.Circe.decodeToFuture[List[ProtocolId]](json)
+    }
+  }
```
[the remark this answers is not in the document]
Comment (ivanopagano) [reply src/main/scala/tech/cryptonomic/conseil/tezos/TezosNodeFetchers.scala]: Everything's correct, apart from a minor correction to the last statements, as the process fetches data and returns the list of _de-serialized_ objects

File: src/main/scala/tech/cryptonomic/conseil/Lorre.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (ivanopagano) [reply src/main/scala/tech/cryptonomic/conseil/Lorre.scala]: @ba205 after looking at the code again, we actually detached from this model somehow.
Even the main block processing cycle fetches from the tezos node and then persists everything in 2 steps: 1. build the queries from TezosDB 2. run them agains the database instance

Right now how voting is stored follows the same pattern. If that is wrong, i.e. if we might want to hide the database usage, we would need to compose the basic operations into the TezosDatabaseOperator as more complex ones.
In any case any measure we might want to pursue seems to more appropriately pertain to a separate PR.

**Pull request merged**